### ISSUE TITLE
ConcurrentModificationException from HealthCheckServiceImpl when an image starting 

### ISSUE THREAD

**Issue by navidsh** [OpenLiberty/open-liberty#9254]
Intermittent `ConcurrentModificationException` happening in Open Liberty based containers running in a Kubernetes environment.

The server is configured to use `mpHealth-1.0` feature and is running a relatively complex app (with DB connection ...).

This error has been seen sporadically since at least `19.0.0.5`.

```
[6/5/19 22:52:11:055 GMT] 00000033 com.ibm.ws.webcontainer.util.ApplicationErrorUtils           E SRVE0777E: Exception thrown by application class 'java.util.HashMap$HashIterator.nextNode:-1'
java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(Unknown Source)
        at java.util.HashMap$KeyIterator.next(Unknown Source)
        at com.ibm.ws.microprofile.health.internal.HealthCheckServiceImpl.performHealthCheck(HealthCheckServiceImpl.java:103)
        at com.ibm.ws.microprofile.health.internal.servlet.HealthCheckServlet.service(HealthCheckServlet.java:53)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255)
        at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743)
        at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440)
        at com.ibm.ws.webcontainer.filter.WebAppFilterChain.invokeTarget(WebAppFilterChain.java:182)
        at com.ibm.ws.webcontainer.filter.WebAppFilterChain.doFilter(WebAppFilterChain.java:93)
        at com.ibm.ws.app.manager.wab.internal.OsgiDirectoryProtectionFilter.doFilter(OsgiDirectoryProtectionFilter.java:90)
        at com.ibm.ws.webcontainer.filter.FilterInstanceWrapper.doFilter(FilterInstanceWrapper.java:201)
        at com.ibm.ws.webcontainer.filter.WebAppFilterChain.doFilter(WebAppFilterChain.java:90)
        at com.ibm.ws.webcontainer.filter.WebAppFilterManager.doFilter(WebAppFilterManager.java:996)
        at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1134)
        at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1005)
        at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
        at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
        at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
        at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1061)
        at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:417)
        at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:376)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:532)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:466)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:331)
        at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:302)
        at com.ibm.ws.channel.ssl.internal.SSLConnectionLink.determineNextChannel(SSLConnectionLink.java:1058)
        at com.ibm.ws.channel.ssl.internal.SSLConnectionLink$MyReadCompletedCallback.complete(SSLConnectionLink.java:644)
        at com.ibm.ws.channel.ssl.internal.SSLReadServiceContext$SSLReadCompletedCallback.complete(SSLReadServiceContext.java:1803)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:503)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:573)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:954)
        at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1043)
        at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:239)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)

[6/5/19 22:52:11:137 GMT] 00000033 com.ibm.ws.logging.internal.impl.IncidentImpl  
```

[messages.log](https://github.com/OpenLiberty/open-liberty/files/3710119/messages.log)
[ffdc_19.06.05_22.52.11.0.log](https://github.com/OpenLiberty/open-liberty/files/3710120/ffdc_19.06.05_22.52.11.0.log)

**Comment by pgunapal** [comment]
@navidsh as discussed over Slack, I had made some updates for mpHealth-1.0 and mpHealth-2.0 in Open Liberty 19.0.0.8, which should resolve this issue. Basically, while the deployed applications/modules map is being iterated over using an iterator, the map is being modified at the same time to add an application, after the application has finished initializing. Can you please ask SVT to try out their scenario using Open Liberty 19.0.0.8 or later and see if the issue is reproducible?

**Comment by navidsh** [comment]
Thanks @pgunapal. I already asked the team to see if they can reproduce it with Open Liberty `19.0.0.8`. Is it possible to backport the fix to `mpHealth-1.0`?

**Comment by pgunapal** [comment]
The fix is for both mpHealth-1.0 and mpHealth-2.0.

**Comment by navidsh** [comment]
We verified that we cannot reproduce the `ConcurrentModificationException` anymore on `19.0.0.9`.

Closing this one.
